loadbalancer/algorithm.go

```diff
@@ -241,10 +241,13 @@ func hash(s string) uint64 {
 // Returns index in hash ring with the closest hash to key's hash
 func (ch *consistentHash) searchRing(key string, skipEndpoint func(int) bool) int {
 	h := hash(key)
-	i := sort.Search(ch.Len(), func(i int) bool { return ch.hashRing[i].hash >= h && !skipEndpoint(ch.hashRing[i].index) })
+	i := sort.Search(ch.Len(), func(i int) bool { return ch.hashRing[i].hash >= h })
 	if i == ch.Len() { // rollover
 		i = 0
 	}
+	for skipEndpoint(ch.hashRing[i].index) {
```

Comment:
The only potential issue I see is that this function is no longer safe if all endpoints are fading, since it would be an infinite loop. I believe this currently can't happen because it's handled at https://github.com/zalando/skipper/blob/24e762b2c17c4d1c94da1cd24dbcc205557028d4/loadbalancer/algorithm.go#L137, but it's not that obvious since it relies on the caller handling this case. So perhaps make the function return the result of the binary search if it does a full loop of the hashring?

Reply:
As for me, the current code seems logical to me. However, not to miss this case next time I've added `TestApplyEndsWhenAllEndpointsAreFading` to check that (LBAlgorithm).Apply finishes for any LBAlgorithm.